client/src/main/java/io/pravega/client/auth/PravegaAuthHandler.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.auth;
+
+import java.util.Map;
+
+/**
+ * Custom authorization/authentication handlers implement this interface.
+ * The implementations are loaded from the classpath using `ServiceLoader` (https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)
+ * Pravega controller also implements this interface through `PravegaDefaultAuthHandler`.
+ *
+ * Each custom auth handler is registered with a unique name.
+ * A client selects its auth handler by setting a grpc header with a name "method". T
+ * his is done by implementing `PravegaCredentials` interface and passing it to client calls.
+ *
+ */
+public interface PravegaAuthHandler {
```

Comment:
It is laudable for authentication to be extensible but I have a serious doubt about the practicality of this design.   It is too simplistic yet also too prescriptive.

My preference is to use [SASL ](https://tools.ietf.org/html/rfc4422) as the framework and to focus on facilitating other *mechanisms*, as opposed to extensions of the *protocol* as `PravegaAuthHandler` intends.

Reply:
`It is too simplistic yet also too prescriptive.` Not sure how this can be countered :)

I tried to figure out how we could use SASL with `grpc` but could not figure out a way to do it. I think `grpc` works with header based auth rather than the traditional connection based auth.

Reply:
I hear you, the Java-GRPC doesn't seem to support channel authentication, just per-call authentication.   So using SASL would be tough.